src/fsharp/TypeChecker.fs

```diff
@@ -8471,6 +8373,14 @@ and TcComputationExpression cenv env overallTy mWhole (interpExpr: Expr) builder
                 let hasBindReturnN = not (isNil (TryFindIntrinsicOrExtensionMethInfo ResultCollectionSettings.AtMostOneResult cenv env bindRange ad bindReturnNName builderTy))
                 if hasBindReturnN && Option.isSome (convertSimpleReturnToExpr varSpace innerComp) then 
                     let consumePat = SynPat.Tuple(false, pats, letPat.Range)
+
+                    // Add the variables to the query variable space, on demand
+                    let varSpace = 
+                        addVarsToVarSpace varSpace (fun _mCustomOp env -> 
+                                use _holder = TemporarilySuspendReportingTypecheckResultsToSink cenv.tcSink
+                                let _, _, vspecs, envinner, _ = TcMatchPattern cenv (NewInferenceType()) env tpenv (consumePat, None) 
+                                vspecs, envinner)
```

Comment:
Is this the core of the change, since the issue was not having access to the variable space?

Reply:
There were two issues.  One was this (it only applied to Bind2Reaturn etc.). The other was in `convertSimpleReturnToExpr`. This function basically detects a thing that can be turned into a BindReturn.  However it wasn't accepting uses of custom operators.

To explain: a custom operator 

    b { part1; op; part2 }

sort of becomes

    b { let! varspace = op (b { part1; return varspace })
        part2 }

The `part1; return varspace` is eligible to become a `BindReturn`.  Previously we were requiring it to be `Bind` then `Return`.